certora/specs/YieldStreams.spec

```diff
@@ -0,0 +1,806 @@
+    env e;
+    uint256 _shares = vault.convertToShares(_principal);
+    require alice != 0 && bob != 0 && alice != bob && bob == e.msg.sender;
+
+    // State changes
+    uint256 streamId = nextStreamId();
+    require streamIdToReceiver(streamId) == bob;
+    require receiverTotalShares(bob) == _shares;
+    require receiverTotalPrincipal(bob) == _principal;
+    require receiverPrincipal(bob, streamId) == _principal;
+    require _shares > 0;
+    require vault.balanceOf(alice) >= _shares;
+    require 1 <= parts;
+    uint256 principalParts;
+    require to_mathint(principalParts) == _principal / parts;
+    uint256 expectedYieldInShares = vault.convertToShares(principalParts);
+    // Add X% profit to vault
+    require to_mathint(vault.balanceOf(currentContract)) == _shares + principalParts;
+
+    uint256 claimed = claimYieldInShares(e, bob);
+
+    assert delta(claimed, expectedYieldInShares) <= 1 &&
+     asset.balanceOf(bob) == 0 &&
+     vault.balanceOf(bob) == claimed &&
+     vault.balanceOf(alice) == 0;
+}
+
+
+/**
+ * @Rule Verify property for the `claimYieldInShares` function for another account
+ * @Category High
+ * @Description Ensures that the `claimYieldInShares` function verifies the correct behavior of the claimYieldInShares function when the receiver (bob) claims the yield in the form of shares to a different account (carol). It sets up a scenario where a yield stream is opened between alice (the streamer) and bob (the receiver), with a specified principal amount. The rule then generates yield by adding a percentage of the principal to the vault's balance. Finally, it invokes the claimYieldInShares function with carol as the recipient and asserts that the claimed yield in shares is correctly calculated and transferred to carol's share balance, while ensuring that the contract's state is updated accordingly.
+ */
+ rule verify_claimYieldInShares_toAnotherAccount(address alice, address bob, address carol, uint256 _principal, uint8 parts) {
+    env e;
+    uint256 _shares = vault.convertToShares(_principal);
+    require alice != 0 && bob != 0 && alice != bob && alice != carol && bob != carol && bob == e.msg.sender;
+
+    // State changes
+    uint256 streamId = nextStreamId();
+    require streamIdToReceiver(streamId) == bob;
+    require receiverTotalShares(bob) == _shares;
+    require receiverTotalPrincipal(bob) == _principal;
+    require receiverPrincipal(bob, streamId) == _principal;
+    require _shares > 0;
+    require vault.balanceOf(alice) >= _shares;
+    require 1 <= parts;
+
+    uint256 principalParts;
+    require to_mathint(principalParts) == _principal / parts;
+
+    // Add X% profit to vault
+    require to_mathint(vault.balanceOf(currentContract)) == _shares + principalParts;
+
+    mathint expectedYieldInShares = vault.convertToShares(principalParts);
+
+    mathint claimed = claimYieldInShares(e, carol);
+
+    assert delta(claimed, expectedYieldInShares) <= 1 &&
```

Comment:
perhaps multiple asserts is easier to read than `&&`, also it would be easier to investigate assertion failure if each assert is about one single postcondition